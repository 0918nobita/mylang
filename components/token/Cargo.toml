[package]
name = "token"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.53"
ast = { path = "../ast" }
serde = { version = "1.0.136", features = ["derive"] }
