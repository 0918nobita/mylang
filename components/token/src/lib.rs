use ast::{
    pos::Pos,
    range::{Locatable, Range},
};

#[derive(Debug)]
pub enum Token {
    I32(Range, i32),
    AddOp(Pos),
    Str(Range, String),
    PrintI32Keyword(Range),
    PrintStrKeyword(Range),
}

impl Locatable for Token {
    fn locate(&self) -> Range {
        match self {
            Token::I32(r, _)
            | Token::Str(r, _)
            | Token::PrintI32Keyword(r)
            | Token::PrintStrKeyword(r) => r.clone(),

            Token::AddOp(p) => p.clone().into(),
        }
    }
}
