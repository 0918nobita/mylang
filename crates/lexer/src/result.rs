//! 字句解析の結果

use mylang_token::{Locatable, Pos, Range};
use thiserror::Error;

/// 字句解析中に発生するエラー
#[derive(Debug, Error)]
pub enum LexErr {
    #[error("Forbidden character: '{1}'")]
    ForbiddenChar(Pos, char),

    #[error("Invalid escape sequence: '{1}'")]
    InvalidEscapeSequence(Pos, char),

    #[error("Invalid keyword: \"{1}\"")]
    InvalidKeyword(Range, String),

    #[error("Missing a closing quote for string literal")]
    MissingClosingQuoteForStr(Pos),
}

impl Locatable for LexErr {
    fn locate(&self) -> Range {
        match self {
            LexErr::ForbiddenChar(pos, _)
            | LexErr::InvalidEscapeSequence(pos, _)
            | LexErr::MissingClosingQuoteForStr(pos) => pos.clone().into(),
            LexErr::InvalidKeyword(range, _) => range.clone(),
        }
    }
}

/// 字句解析の結果
pub type LexResult<T> = Result<T, LexErr>;
