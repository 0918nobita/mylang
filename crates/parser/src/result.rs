use mylang_token::{Locatable, Pos, Range};
use thiserror::Error;

/// 構文解析中に発生するエラー
#[derive(Debug, Error)]
pub enum ParseErr {
    #[error("Term expected, but not found")]
    TermExpected(Pos),

    #[error("Either print_int or print_str expected, but not found")]
    KeywordExpected(Range),
}

impl Locatable for ParseErr {
    fn locate(&self) -> Range {
        match self {
            ParseErr::TermExpected(pos) => pos.clone().into(),

            ParseErr::KeywordExpected(range) => range.clone(),
        }
    }
}

pub type ParseResult<T> = Result<T, ParseErr>;
