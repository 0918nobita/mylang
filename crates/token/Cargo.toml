[package]
name = "mylang_token"
version = "0.1.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.143", features = ["derive"] }
