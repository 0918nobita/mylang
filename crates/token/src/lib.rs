//! 字句解析で取り出されるトークンとその付帯情報の定義

mod locatable;
mod pos;
mod range;

use serde::{Deserialize, Serialize};

pub use locatable::Locatable;
pub use pos::Pos;
pub use range::Range;

/// [`Pos`] を簡単に生成するためのマクロ
#[macro_export]
macro_rules! pos {
    ($line:expr ; $char:expr) => {
        mylang_token::Pos::new($line, $char)
    };
}

/// [`Range`] を簡単に生成するためのマクロ
#[macro_export]
macro_rules! range {
    ($start_line:expr ; $start_char:expr, $end_line:expr ; $end_char:expr) => {
        mylang_token::Range::new(
            mylang_token::pos!($start_line;$start_char),
            mylang_token::pos!($end_line;$end_char),
        )
    };
    ($start:expr, $end:expr) => {
        mylang_token::Range::new($start, $end)
    };
}

#[derive(Debug, Serialize, Deserialize)]
pub enum KeywordKind {
    Let,
    In,
    PrintI32,
    PrintStr,
}

impl KeywordKind {
    pub fn parse(str: &str) -> Option<Self> {
        match str {
            "let" => Some(Self::Let),
            "in" => Some(Self::In),
            "print_int" => Some(Self::PrintI32),
            "print_str" => Some(Self::PrintStr),
            _ => None,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub enum Token {
    I32(Range, i32),
    AddOp(Pos),
    Str(Range, String),
    Keyword(Range, KeywordKind),
    Ident(Range, String),
    Newline(Pos),
}

impl Locatable for Token {
    fn locate(&self) -> Range {
        match self {
            Self::I32(r, _) | Self::Str(r, _) | Self::Keyword(r, _) | Self::Ident(r, _) => {
                r.clone()
            }

            Self::AddOp(p) | Self::Newline(p) => p.clone().into(),
        }
    }
}
