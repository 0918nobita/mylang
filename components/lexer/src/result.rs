//! 字句解析の結果

use ast::{pos::Pos, range::Range};
use thiserror::Error;
use token::Token;

/// 字句解析中に発生するエラー
#[derive(Debug, Error)]
pub enum LexErr {
    #[error("({0}) Forbidden character: '{1}'")]
    ForbiddenChar(Pos, char),

    #[error("({0}) Invalid escape sequence: '{1}'")]
    InvalidEscapeSequence(Pos, char),

    #[error("({0}) Invalid keyword: \"{1}\"")]
    InvalidKeyword(Range, String),

    #[error("({0}) Missing a closing quote for string literal")]
    MissingClosingQuoteForStr(Pos),
}

/// 字句解析中に得られたトークン、または発生したエラー
pub type LexResult = Result<Token, LexErr>;
