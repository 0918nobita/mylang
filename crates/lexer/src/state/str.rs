use token::{Pos, Range, Token};

use crate::result::LexErr;

#[derive(Clone, Debug)]
pub struct StrState {
    start: Pos,
    pub end: Pos,
    pub escape: bool,
    acc: String,
}

impl StrState {
    pub fn new(pos: Pos) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
            escape: false,
            acc: String::new(),
        }
    }

    pub fn try_append_char(&self, pos: Pos, c: char) -> Result<Self, LexErr> {
        match (self.escape, c) {
            (_, '\n') => Err(LexErr::ForbiddenChar(pos, c)),

            (true, c @ ('\\' | 'n')) => Ok(Self {
                start: self.start.clone(),
                end: pos,
                escape: false,
                acc: format!("{}{c}", self.acc),
            }),

            (true, _) => Err(LexErr::InvalidEscapeSequence(pos, c)),

            (false, '\\') => Ok(Self {
                start: self.start.clone(),
                end: pos,
                escape: true,
                acc: self.acc.clone(),
            }),

            (false, c) => Ok(Self {
                start: self.start.clone(),
                end: pos,
                escape: false,
                acc: format!("{}{c}", self.acc),
            }),
        }
    }

    pub fn tokenize(&self) -> Token {
        if self.escape {
            panic!("Illigal state")
        }
        Token::Str(
            Range::new(self.start.clone(), self.end.clone()),
            self.acc.clone(),
        )
    }
}
