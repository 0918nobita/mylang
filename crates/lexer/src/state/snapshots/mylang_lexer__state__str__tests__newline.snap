---
source: crates/lexer/src/state/str.rs
assertion_line: 73
expression: state

---
ForbiddenChar(
    0:0,
    '\n',
)
